File: core/src/main/java/org/berndpruenster/jtor/mgmt/TorEventHandler.java
/*
Copyright (C) 2011-2014 Sublime Software Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */

/*
 Copyright (c) 2016, 2017 Bernd Prünster
 Copyright (c) 2014-2015 Microsoft Open Technologies, Inc.
 All Rights Reserved
 Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
 License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0

 THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 INCLUDING WITHOUT LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
 MERCHANTABLITY OR NON-INFRINGEMENT.

 See the Apache 2 License for the specific language governing permissions and limitations under the License.
 */

package org.berndpruenster.jtor.mgmt;

import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import org.berndpruenster.jtor.socket.HiddenServiceSocket;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import net.freehaven.tor.control.EventHandler;

/**
 * Logs the data we get from notifications from the Tor OP. This is really just
 * meant for debugging.
 */
public class TorEventHandler implements EventHandler {

  private static final String                                 UPLOADED = "UPLOADED";
  private static final String                                 HS_DESC  = "HS_DESC";
  private static final Logger                                 LOG      = LoggerFactory.getLogger(TorEventHandler.class);

  private final Map<String, HiddenServiceSocket>              socketMap;
  private final Map<String, List<HiddenServiceReadyListener>> listenerMap;

  TorEventHandler() {
    this.socketMap = new HashMap<>();
    this.listenerMap = new HashMap<>();
  }

  public void attachReadyListeners(final HiddenServiceSocket hs, final List<HiddenServiceReadyListener> listeners) {
    synchronized (socketMap) {
      socketMap.put(hs.getServiceName(), hs);
      listenerMap.put(hs.getServiceName(), listeners);
    }
  }

  @Override
  public void circuitStatus(final String status, final String id, final String path) {
    final String msg = "CircuitStatus: " + id + " " + status + ", " + path;
    LOG.debug(msg);
  }

  @Override
  public void streamStatus(final String status, final String id, final String target) {
    final String msg = "streamStatus: status: " + status + ", id: " + id + ", target: " + target;
    LOG.debug(msg);

  }

  @Override
  public void orConnStatus(final String status, final String orName) {
    final String msg = "OR connection: status: " + status + ", orName: " + orName;
    LOG.debug(msg);
  }

  @Override
  public void bandwidthUsed(final long read, final long written) {
    LOG.debug("bandwidthUsed: read: " + read + ", written: " + written);
  }

  @Override
  public void newDescriptors(final List<String> orList) {
    final Iterator<String> iterator = orList.iterator();
    final StringBuilder stringBuilder = new StringBuilder();
    while (iterator.hasNext()) {
      stringBuilder.append(iterator.next());
    }
    final String msg = "newDescriptors: " + stringBuilder.toString();
    LOG.debug(msg);

  }

  @Override
  public void message(final String severity, final String msg) {
    final String msg2 = "message: severity: " + severity + ", msg: " + msg;
    LOG.debug(msg2);
  }

  @Override
  public void unrecognized(final String type, final String msg) {
    final String msg2 = "unrecognized: type: " + type + ", msg: " + msg;
    LOG.debug(msg2);
    if (type.equals(HS_DESC) && msg.startsWith(UPLOADED)) {
      final String hiddenServiceID = msg.split(" ")[1] + ".onion";
      synchronized (socketMap) {
        final HiddenServiceSocket hs = socketMap.get(hiddenServiceID);
        if (hs == null) {
          return;
        }
        LOG.info("Hidden Service " + hs + " is ready");
        for (final HiddenServiceReadyListener listener : listenerMap.get(hiddenServiceID)) {
          listener.onReady(hs);
        }
        socketMap.remove(hiddenServiceID);
        listenerMap.remove(hiddenServiceID);
      }
    }
  }

}

File: tor/src/main/kotlin/org/berndpruenster/netlayer/tor/TorEventHandler.kt
/*
Copyright (c) 2016, 2017 Bernd Prünster
This file is part of of the unofficial Java-Tor-bindings.

Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the
European Commission - subsequent versions of the EUPL (the "Licence"); You may
not use this work except in compliance with the Licence. You may obtain a copy
of the Licence at: http://joinup.ec.europa.eu/software/page/eupl

Unless required by applicable law or agreed to in writing, software distributed
under the Licence is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the Licence for the
specific language governing permissions and limitations under the Licence.

This project includes components developed by third parties and provided under
various open source licenses (www.opensource.org).


 Copyright (c) 2014-2015 Microsoft Open Technologies, Inc.
 Copyright (C) 2011-2014 Sublime Software Ltd

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/

package org.berndpruenster.netlayer.tor

import net.freehaven.tor.control.EventHandler
import kotlin.concurrent.thread

private const val UPLOADED = "UPLOADED"
private const val RECEIVED = "RECEIVED"

/**
 * Manages and triggers the ready-callbacks
 */
class TorEventHandler : EventHandler {

    private val listenerMap = HashMap<String, () -> Unit>()

    class HiddenServiceSocketList(private val hs: HiddenServiceSocket, private val listeners: List<(socket: HiddenServiceSocket) -> Unit>) : () -> Unit {
        override fun invoke() {
            listeners.forEach {
                thread{it(hs)}
            }
        }
    }

    fun attachReadyListeners(hs: HiddenServiceSocket, listeners: List<(socket: HiddenServiceSocket) -> Unit>) {
        synchronized(listenerMap) {
            listenerMap.put(hs.socketAddress.serviceName, HiddenServiceSocketList(hs, listeners))
        }
    }

    fun attachHSReadyListener(serviceName: String, listener: () -> Unit) {
        synchronized(listenerMap) {
            listenerMap.put(serviceName, listener)
        }
    }

    override fun circuitStatus(status: String, id: String, path: String) {
        val msg = "CircuitStatus: $id $status $path"
        logger?.debug(msg)
    }

    override fun streamStatus(status: String, id: String, target: String) {
        val msg = "streamStatus: status: $status $id: , target: $target"
        logger?.debug(msg)

    }

    override fun orConnStatus(status: String, orName: String) {
        val msg = "OR connection: status: $status, orName: $orName"
        logger?.debug(msg)
    }

    override fun bandwidthUsed(read: Long, written: Long) {
        logger?.debug("bandwidthUsed: read: $read , written: $written")
    }

    override fun newDescriptors(orList: List<String>) {

        val stringBuilder = StringBuilder("newDescriptors: ")

        orList.forEach {
            stringBuilder.append(it)
        }
        logger?.debug(stringBuilder.toString())

    }

    override fun message(severity: String, msg: String) {
        val msg2 = "message: severity: $severity , msg: $msg"
        logger?.trace(msg2)
    }

    override fun hiddenServiceEvent(type: String, msg: String) {
        logger?.debug("hiddenService: HS_DESC $msg")
        when(type) {
            UPLOADED -> {
                val hiddenServiceID = "${msg.split(" ")[1]}.onion"
                synchronized(listenerMap) {
                    logger?.info("Hidden Service $hiddenServiceID has been announced to the Tor network.")
                    listenerMap[hiddenServiceID]?.run {thread(block = this)}
                    listenerMap.remove(hiddenServiceID)
                }
            }
        }
    }

    override fun hiddenServiceFailedEvent(reason: String, msg: String) {
        logger?.debug("hiddenService: HS_DESC $msg")
    }

    override fun hiddenServiceDescriptor(descriptorId: String, descriptor: String, msg: String) {
        logger?.debug("hiddenService: HS_DESC_CONTENT $descriptorId $descriptor, as in $msg")
    }

    override fun timeout() {
        logger?.debug("The control connection to tor did not provide a response within one minute of waiting.")
    }

    override fun unrecognized(type: String, msg: String) {
        val msg2 = "unrecognized: current: $type , $msg: msg"
        logger?.debug(msg2)
    }

}
